wpilibc/src/main/native/include/frc/estimator/KalmanFilter.h

```diff
@@ -0,0 +1,230 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2018-2020 FIRST. All Rights Reserved.                        */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <array>
+#include <cmath>
+
+#include <Eigen/Core>
+#include <drake/math/discrete_algebraic_riccati_equation.h>
+#include <units/time.h>
+
+#include "frc/StateSpaceUtil.h"
+#include "frc/system/Discretization.h"
+#include "frc/system/LinearSystem.h"
+
+namespace frc {
+
+/**
+ * Luenberger observers combine predictions from a model and measurements to
+ * give an estimate of the true system state.
+ *
+ * Luenberger observers use an L gain matrix to determine whether to trust the
+ * model or measurements more. Kalman filter theory uses statistics to compute
+ * an optimal L gain (alternatively called the Kalman gain, K) which minimizes
+ * the sum of squares error in the state estimate.
+ *
+ * Luenberger observers run the prediction and correction steps simultaneously
+ * while Kalman filters run them sequentially. To implement a discrete-time
+ * Kalman filter as a Luenberger observer, use the following mapping:
+ * <pre>C = H, L = A * K</pre>
+ * (H is the measurement matrix).
+ *
+ * For more on the underlying math, read
+ * https://file.tavsys.net/control/state-space-guide.pdf.
+ */
+template <int States, int Inputs, int Outputs>
+class KalmanFilter {
```

Comment:
> For your KF, I'd suggest only supporting constant dt, which means constant A, B, C, D matrices.  You can then pick to either set an initial state and distribution, or to run it at steady state all the time.  We on 971 solve for the steady state kalman gain and covariances and just use those.

-- Austin Schuh

I didn't see a difference between this impl and one with a constant dt when computing the Kalman gain, so I agree with this simplification. However, it breaks the latency compensators, so we should merge what we have, imo, and do the change later.

I have a steady-state KF done in https://github.com/calcmogul/allwpilib/tree/state-space-v2-steady-state-kf.